File: src/models/Listing.js
const mongoose = require('mongoose');

const ListingSchema = new mongoose.Schema({
  hostId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  title: {
    en: {
      type: String,
      required: [true, 'Please provide a title in English'],
      trim: true,
      maxlength: 100,
    },
    sw: {
      type: String,
      required: [true, 'Please provide a title in Swahili'],
      trim: true,
      maxlength: 100,
    },
  },
  description: {
    en: {
      type: String,
      required: [true, 'Please provide a description in English'],
      maxlength: 2000,
    },
    sw: {
      type: String,
      required: [true, 'Please provide a description in Swahili'],
      maxlength: 2000,
    },
  },
  propertyType: {
    type: String,
    required: true,
    enum: [
      'apartment', 'house', 'villa', 'guesthouse',
      'hotel', 'resort', 'cottage', 'bungalow', 'studio'
    ],
  },
  location: {
    address: {
      type: String,
      required: true,
    },
    region: {
      type: String,
      required: true,
      enum: [
        'Dar es Salaam', 'Arusha', 'Dodoma', 'Mwanza',
        'Mbeya', 'Tanga', 'Zanzibar', 'Kilimanjaro',
        'Morogoro', 'Pwani', 'Other'
      ],
    },
    city: {
      type: String,
      required: true,
    },
    district: String,
    coordinates: {
      type: {
        type: String,
        enum: ['Point'],
        required: true,
      },
      coordinates: {
        type: [Number],
        required: true,
      },
    },
    nearbyLandmarks: [String],
  },
  pricing: {
    basePrice: {
      type: Number,
      required: [true, 'Please provide a base price'],
      min: 0,
    },
    currency: {
      type: String,
      default: 'TZS',
    },
    cleaningFee: {
      type: Number,
      default: 0,
    },
    serviceFee: {
      type: Number,
      default: 0,
    },
    weeklyDiscount: {
      type: Number,
      default: 0,
      min: 0,
      max: 100,
    },
    monthlyDiscount: {
      type: Number,
      default: 0,
      min: 0,
      max: 100,
    },
    extraGuestFee: {
      type: Number,
      default: 0,
    },
  },
  capacity: {
    guests: {
      type: Number,
      required: true,
      min: 1,
    },
    bedrooms: {
      type: Number,
      required: true,
      min: 0,
    },
    beds: {
      type: Number,
      required: true,
      min: 1,
    },
    bathrooms: {
      type: Number,
      required: true,
      min: 1,
    },
  },
  amenities: [{
    type: String,
    enum: [
      'wifi', 'parking', 'kitchen', 'air_conditioning', 'heating',
      'tv', 'pool', 'gym', 'security', 'generator', 'water_tank',
      'washer', 'dryer', 'workspace', 'breakfast', 'pet_friendly',
      'family_friendly', 'accessible', 'smoking_allowed', 'mosquito_nets'
    ],
  }],
  images: [{
    url: {
      type: String,
      required: true,
    },
    caption: String,
    order: {
      type: Number,
      default: 0,
    },
  }],
  houseRules: {
    checkIn: {
      type: String,
      default: '14:00',
    },
    checkOut: {
      type: String,
      default: '11:00',
    },
    customRules: {
      en: [String],
      sw: [String],
    },
  },
  availability: {
    minNights: {
      type: Number,
      default: 1,
      min: 1,
    },
    maxNights: {
      type: Number,
      default: 365,
    },
    instantBooking: {
      type: Boolean,
      default: false,
    },
  },
  blockedDates: {
    type: [{
      start: {
        type: Date,
        required: true,
      },
      end: {
        type: Date,
        required: true,
      },
      reason: {
        type: String,
        trim: true,
      },
      createdAt: {
        type: Date,
        default: Date.now,
      },
      createdBy: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
      },
    }],
    default: [],
  },
  rating: {
    average: {
      type: Number,
      default: 0,
      min: 0,
      max: 5,
    },
    count: {
      type: Number,
      default: 0,
    },
    breakdown: {
      cleanliness: { type: Number, default: 0 },
      accuracy: { type: Number, default: 0 },
      communication: { type: Number, default: 0 },
      location: { type: Number, default: 0 },
      checkIn: { type: Number, default: 0 },
      value: { type: Number, default: 0 },
    },
  },
  status: {
    type: String,
    enum: ['draft', 'active', 'inactive', 'suspended'],
    default: 'draft',
  },
  views: {
    type: Number,
    default: 0,
  },
  bookingsCount: {
    type: Number,
    default: 0,
  },
  featured: {
    type: Boolean,
    default: false,
  },
  verifiedByAdmin: {
    type: Boolean,
    default: false,
  },
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true },
});

// Create geospatial index
ListingSchema.index({ 'location.coordinates': '2dsphere' });

// Create text index for search
ListingSchema.index({
  'title.en': 'text',
  'title.sw': 'text',
  'description.en': 'text',
  'description.sw': 'text',
});

// Additional indexes
ListingSchema.index({ hostId: 1 });
ListingSchema.index({ 'location.region': 1 });
ListingSchema.index({ propertyType: 1 });
ListingSchema.index({ 'pricing.basePrice': 1 });
ListingSchema.index({ 'rating.average': -1 });
ListingSchema.index({ status: 1 });
ListingSchema.index({ createdAt: -1 });

// Populate host info
ListingSchema.pre(/^find/, function(next) {
  this.populate({
    path: 'hostId',
    select: 'profile.firstName profile.lastName profile.profilePicture rating isEmailVerified isPhoneVerified',
  });
  next();
});

module.exports = mongoose.model('Listing', ListingSchema);

File: mobile/lib/core/models/listing.dart
class Listing {
  final String id;
  final String hostId;
  final Map<String, dynamic>? host;  // Full host object data
  final LocalizedText title;
  final LocalizedText description;
  final String propertyType;
  final Location location;
  final Pricing pricing;
  final Capacity capacity;
  final List<String> amenities;
  final List<ListingImage> images;
  final Rating rating;
  final String status;
  final int views;
  final bool featured;
  
  Listing({
    required this.id,
    required this.hostId,
    this.host,
    required this.title,
    required this.description,
    required this.propertyType,
    required this.location,
    required this.pricing,
    required this.capacity,
    required this.amenities,
    required this.images,
    required this.rating,
    required this.status,
    required this.views,
    required this.featured,
  });
  
  factory Listing.fromJson(Map<String, dynamic> json) {
    return Listing(
      id: json['_id'] ?? json['id'],
      hostId: json['hostId'] is String ? json['hostId'] : (json['hostId']?['_id'] ?? ''),
      host: json['hostId'] is Map<String, dynamic> ? Map<String, dynamic>.from(json['hostId']) : null,
      title: LocalizedText.fromJson(json['title']),
      description: LocalizedText.fromJson(json['description']),
      propertyType: json['propertyType'],
      location: Location.fromJson(json['location']),
      pricing: Pricing.fromJson(json['pricing']),
      capacity: Capacity.fromJson(json['capacity']),
      amenities: List<String>.from(json['amenities'] ?? []),
      images: (json['images'] as List)
          .map((img) => ListingImage.fromJson(img))
          .toList(),
      rating: Rating.fromJson(json['rating']),
      status: json['status'],
      views: json['views'] ?? 0,
      featured: json['featured'] ?? false,
    );
  }
  
  String get mainImage => images.isNotEmpty ? images.first.url : '';
}

class LocalizedText {
  final String en;
  final String sw;
  
  LocalizedText({required this.en, required this.sw});
  
  factory LocalizedText.fromJson(Map<String, dynamic> json) {
    return LocalizedText(
      en: json['en'] ?? '',
      sw: json['sw'] ?? '',
    );
  }
  
  String get(String locale) => locale == 'sw' ? sw : en;
}

class Location {
  final String address;
  final String region;
  final String city;
  final double latitude;
  final double longitude;
  
  Location({
    required this.address,
    required this.region,
    required this.city,
    required this.latitude,
    required this.longitude,
  });
  
  factory Location.fromJson(Map<String, dynamic> json) {
    final coords = json['coordinates']['coordinates'] as List;
    return Location(
      address: json['address'],
      region: json['region'],
      city: json['city'],
      longitude: coords[0].toDouble(),
      latitude: coords[1].toDouble(),
    );
  }
}

class Pricing {
  final double basePrice;
  final String currency;
  final double cleaningFee;
  final double serviceFee;
  
  Pricing({
    required this.basePrice,
    required this.currency,
    required this.cleaningFee,
    required this.serviceFee,
  });
  
  factory Pricing.fromJson(Map<String, dynamic> json) {
    return Pricing(
      basePrice: (json['basePrice'] ?? 0).toDouble(),
      currency: json['currency'] ?? 'TZS',
      cleaningFee: (json['cleaningFee'] ?? 0).toDouble(),
      serviceFee: (json['serviceFee'] ?? 0).toDouble(),
    );
  }
}

class Capacity {
  final int guests;
  final int bedrooms;
  final int beds;
  final int bathrooms;
  
  Capacity({
    required this.guests,
    required this.bedrooms,
    required this.beds,
    required this.bathrooms,
  });
  
  factory Capacity.fromJson(Map<String, dynamic> json) {
    return Capacity(
      guests: json['guests'] ?? 0,
      bedrooms: json['bedrooms'] ?? 0,
      beds: json['beds'] ?? 0,
      bathrooms: json['bathrooms'] ?? 0,
    );
  }
}

class ListingImage {
  final String url;
  final String? caption;
  final int order;
  
  ListingImage({
    required this.url,
    this.caption,
    required this.order,
  });
  
  factory ListingImage.fromJson(Map<String, dynamic> json) {
    return ListingImage(
      url: json['url'],
      caption: json['caption'],
      order: json['order'] ?? 0,
    );
  }
}

class Rating {
  final double average;
  final int count;
  final RatingBreakdown breakdown;
  
  Rating({
    required this.average, 
    required this.count,
    required this.breakdown,
  });
  
  factory Rating.fromJson(Map<String, dynamic> json) {
    return Rating(
      average: (json['average'] ?? 0).toDouble(),
      count: json['count'] ?? 0,
      breakdown: json['breakdown'] != null 
          ? RatingBreakdown.fromJson(json['breakdown'])
          : RatingBreakdown(
              cleanliness: 0,
              accuracy: 0,
              communication: 0,
              location: 0,
              checkIn: 0,
              value: 0,
            ),
    );
  }
}

class RatingBreakdown {
  final double cleanliness;
  final double accuracy;
  final double communication;
  final double location;
  final double checkIn;
  final double value;
  
  RatingBreakdown({
    required this.cleanliness,
    required this.accuracy,
    required this.communication,
    required this.location,
    required this.checkIn,
    required this.value,
  });
  
  factory RatingBreakdown.fromJson(Map<String, dynamic> json) {
    return RatingBreakdown(
      cleanliness: (json['cleanliness'] ?? 0).toDouble(),
      accuracy: (json['accuracy'] ?? 0).toDouble(),
      communication: (json['communication'] ?? 0).toDouble(),
      location: (json['location'] ?? 0).toDouble(),
      checkIn: (json['checkIn'] ?? 0).toDouble(),
      value: (json['value'] ?? 0).toDouble(),
    );
  }
}





